pulsar-transaction/coordinator/src/main/java/org/apache/pulsar/transaction/coordinator/impl/TxnLogBufferedWriter.java

```diff
@@ -0,0 +1,453 @@
+            return;
+        }
+    }
+
+    private void doFlush(){
+        // Combine data.
+        ByteBuf prefix = PulsarByteBufAllocator.DEFAULT.buffer(4);
+        prefix.writeShort(BATCHED_ENTRY_DATA_PREFIX_MAGIC_NUMBER);
+        prefix.writeShort(BATCHED_ENTRY_DATA_PREFIX_VERSION);
+        ByteBuf actualContent = this.dataSerializer.serialize(this.dataArray);
+        ByteBuf pairByteBuf = Unpooled.wrappedUnmodifiableBuffer(prefix, actualContent);
+        FlushContext<T> flushContext = FlushContext.newInstance(this.asyncAddArgsList);
+        // Clear buffers.
+        this.dataArray.clear();
+        this.asyncAddArgsList = null;
+        this.bytesSize = 0;
+        // Flush.
+        managedLedger.asyncAddEntry(pairByteBuf, this, flushContext);
+    }
+
+    /**
+     * see {@link AsyncCallbacks.AddEntryCallback#addComplete(Position, ByteBuf, Object)}.
+     */
+    @Override
+    public void addComplete(Position position, ByteBuf entryData, Object ctx) {
+        final FlushContext<T> flushContext = (FlushContext<T>) ctx;
+        try {
+            final int batchSize = flushContext.asyncAddArgsList.size();
+            for (int batchIndex = 0; batchIndex < batchSize; batchIndex++) {
+                final AsyncAddArgs asyncAddArgs = flushContext.asyncAddArgsList.get(batchIndex);
+                BitSetRecyclable bitSetRecyclable = BitSetRecyclable.create();
+                bitSetRecyclable.set(batchIndex);
+                long[] ackSet = bitSetRecyclable.toLongArray();
+                bitSetRecyclable.recycle();
+                final TxnBatchedPositionImpl txBatchedPosition = new TxnBatchedPositionImpl(position, batchSize,
+                        batchIndex, ackSet);
+                // Because this task already running at ordered task, so just "run".
+                SafeRunnable.safeRun(() -> {
+                    asyncAddArgs.callback.addComplete(txBatchedPosition, asyncAddArgs.ctx);
+                }).run();
+            }
+        } finally {
+            entryData.release();
+            flushContext.recycle();
+        }
+    }
+
+    /**
+     * see {@link AsyncCallbacks.AddEntryCallback#addFailed(ManagedLedgerException, Object)}.
+     */
+    @Override
+    public void addFailed(ManagedLedgerException exception, Object ctx) {
+        final FlushContext<T> flushContext = (FlushContext<T>) ctx;
+        try {
+            for (int i = 0; i < flushContext.asyncAddArgsList.size(); i++) {
+                final AsyncAddArgs asyncAddArgs = flushContext.asyncAddArgsList.get(i);
+                // Because this task already running at ordered task, so just "run".
+                SafeRunnable.safeRun(() -> asyncAddArgs.callback.addFailed(exception, asyncAddArgs.ctx)).run();
+            }
+        } finally {
+            flushContext.recycle();
+        }
+    }
+
+    /**
+     * Release resources and executing tasks.
+     */
+    @Override
+    public void close() {
+        if (this.scheduledFuture != null) {
```

Comment:
We should also failed the pending writes, clear the pending writes/

Reply:
Good suggestion.

Reply:
I have change the logic:  

1. Force flush once before actual closed. 
2. Change state of buffered writer to closed, avoid new requests coming in.

> Why force flush once before actual closed ? 

This ensures that the logic is always consistent, since it is possible that asynchronous timed refreshes are also being performed when closing.  If ML has been closed, pending requests will be triggered by ML to fail.